### PR TITLE
Make placement options in marker and hash line symbol layers checkboxes instead of mutually exclusive buttons

### LINKED ISSUE #18433: Marker placement options for marker line should be checkboxes
Author Name: **Anita Graser** (@anitagraser)
Original Redmine Issue: [9922](https://issues.qgis.org/issues/9922)

Redmine category:symbology

---
Currently, all options are radio buttons. The user cannot - in one symbol layer - put the same symbol e.g. on the first and the last vertex. The symbol layer has to be duplicated to achieve this. I think that's unnecessary and more difficult to maintain. Therefore, I suggest to use checkboxes instead of radio buttons for the options "on first vertex", "on last vertex" and "on central point".

### PR DESCRIPTION
**Pull request by nyalldawson**
Makes it considerably quicker to create a marker line with the markers on the first AND last vertex, amongst other things...

Fixes #18433

Sponsored by North Road, thanks to SLYR

### CODE REVIEW

**Comment by 3nids** [comment]
nice!

File: src/core/qgis.h [hunk 1]
```diff
@@ -1058,15 +1058,16 @@ class CORE_EXPORT Qgis
...
-      CentralPoint, //!< Place symbols at the mid point of the line
-      CurvePoint, //!< Place symbols at every virtual curve point in the line (used when rendering curved geometry types only)
-      SegmentCenter, //!< Place symbols at the center of every line segment
+      Interval = 1 << 0, //!< Place symbols at regular intervals
+      Vertex = 1 << 1, //!< Place symbols on every vertex in the line
+      LastVertex = 1 << 2, //!< Place symbols on the last vertex in the line
+      FirstVertex = 1 << 3, //!< Place symbols on the first vertex in the line
+      CentralPoint = 1 << 4, //!< Place symbols at the mid point of the line
+      CurvePoint = 1 << 5, //!< Place symbols at every virtual curve point in the line (used when rendering curved geometry types only)
+      SegmentCenter = 1 << 6, //!< Place symbols at the center of every line segment
```
Comment (3nids) [inline src/core/qgis.h:1067]: is this worth a note in API break?

File: src/core/symbology/qgslinesymbollayer.cpp [hunk 2]
```diff
@@ -1454,58 +1471,30 @@ QVariantMap QgsTemplatedLineSymbolLayerBase::properties() const
...
-      map[QStringLiteral( "placement" )] = QStringLiteral( "segmentcenter" );
-      break;
-  }
+  map[QStringLiteral( "placement_vertex" )] = static_cast< bool >( mPlacements & Qgis::MarkerLinePlacement::Vertex );
+  map[QStringLiteral( "placement_last_vertex" )] = static_cast< bool >( mPlacements & Qgis::MarkerLinePlacement::LastVertex );
+  map[QStringLiteral( "placement_first_vertex" )] = static_cast< bool >( mPlacements & Qgis::MarkerLinePlacement::FirstVertex );
+  map[QStringLiteral( "placement_central_point" )] = static_cast< bool >( mPlacements & Qgis::MarkerLinePlacement::CentralPoint );
+  map[QStringLiteral( "placement_curve_point" )] = static_cast< bool >( mPlacements & Qgis::MarkerLinePlacement::CurvePoint );
+  map[QStringLiteral( "placement_interval" )] = static_cast< bool >( mPlacements & Qgis::MarkerLinePlacement::Interval );
+  map[QStringLiteral( "placement_segment_center" )] = static_cast< bool >( mPlacements & Qgis::MarkerLinePlacement::SegmentCenter );
```
Comment (3nids) [inline src/core/symbology/qgslinesymbollayer.cpp:1480]:
```suggestion
map[QStringLiteral( "placements" )] = qgsFlagValueToKeys<Qgis::MarkerLinePlacements>( mPlacements );
```

File: src/core/symbology/qgslinesymbollayer.cpp [hunk 3]
```diff
@@ -1573,19 +1562,38 @@ void QgsTemplatedLineSymbolLayerBase::setCommonProperties( QgsTemplatedLineSymbo
...
+    if ( properties.value( QStringLiteral( "placement_first_vertex" ) ).toBool() )
+      placements |= Qgis::MarkerLinePlacement::FirstVertex;
+    if ( properties.value( QStringLiteral( "placement_central_point" ) ).toBool() )
+      placements |= Qgis::MarkerLinePlacement::CentralPoint;
+    if ( properties.value( QStringLiteral( "placement_curve_point" ) ).toBool() )
+      placements |= Qgis::MarkerLinePlacement::CurvePoint;
+    if ( properties.value( QStringLiteral( "placement_interval" ) ).toBool() )
+      placements |= Qgis::MarkerLinePlacement::Interval;
+    if ( properties.value( QStringLiteral( "placement_segment_center" ) ).toBool() )
+      placements |= Qgis::MarkerLinePlacement::SegmentCenter;
```
Comment (3nids) [inline src/core/symbology/qgslinesymbollayer.cpp:1595]:
```suggestion
    Qgis::MarkerLinePlacements placements = qgsFlagKeysToValue<Qgis::MarkerLinePlacements>( properties.value( QStringLiteral( "placements" ) ).toString() );
```

**Review by 3nids — APPROVED** [review]

File: src/core/symbology/qgslinesymbollayer.cpp (hunk 3 above)
Comment (nyalldawson) [reply src/core/symbology/qgslinesymbollayer.cpp:1595]: How does that store multiple values?

Comment (3nids) [reply src/core/symbology/qgslinesymbollayer.cpp:1595]: https://github.com/qgis/QGIS/blob/release-3_22/tests/src/core/testqgis.cpp#L455

Comment (3nids) [reply src/core/symbology/qgslinesymbollayer.cpp:1595]: this is used for instance to save the configuration flags in vector layer

File: src/core/qgis.h (hunk 1 above)
Comment (nyalldawson) [reply src/core/qgis.h:1067]: I don't think it's necessary -- I can't think of any situation where someone would be using these as a an integer value

File: src/core/qgis.h [hunk 4]
```diff
@@ -1058,15 +1058,16 @@ class CORE_EXPORT Qgis
...
     };
     Q_ENUM( MarkerLinePlacement )
+    Q_DECLARE_FLAGS( MarkerLinePlacements, MarkerLinePlacement )
```
Comment (3nids) [inline src/core/qgis.h:1070]:
```suggestion
    Q_DECLARE_FLAGS( MarkerLinePlacements, MarkerLinePlacement )
    Q_FLAG( MarkerLinePlacements )
```

**Pull request merged**